### ISSUE TITLE
MLPRegression with some categorical columns returns NaN

### ISSUE THREAD

**Issue by midikiman** [chen0040/java-ann-mlp#1]
Trying to train a regressor on a data set with both numeric and categorical columns. The data rows have the expected one-hot category settings, but after fitting transform() always returns NaN, even for rows in the training set. (If the category columns are left out of the frame,  the problem doesn't occur, but the fit isn't good.)

The data set is the training data for https://www.kaggle.com/c/nomad2018-predict-transparent-conductors, where the id column is omitted and the following two columns are categories, if that helps debugging.

**Comment by chen0040** [comment]
Hi @midikiman thanks for reporting the issue. as i don't have your sample source codes, i am not sure what actually happened in your case. at the meantime, I have downloaded transparent-conductors train.csv file and test run with some sample codes for regression (please see my sample code below), the transform() after fitting seems to be working for the categorical and numeric input columns as well as one output numeric column. It is possible that the way you are using some api exposes some bug in the library, but without your sample code, it is not clear on what exactly goes wrong, therefore it will be great if you can share with me your sample code so that i can further debug

`

        InputStream inputStream = FileUtils.getResource("train.csv");
        DataFrame dataFrame = DataQuery.csv(",").from(inputStream).skipRows(1)
                .selectColumn(1)
                .asCategory()
                .asInput("spacegroup")
                .selectColumn(2)
                .asNumeric()
                .asInput("number_of_total_atoms")
                .selectColumn(12)
                .asNumeric()
                .asOutput("formation_energy_ev_natom")
                .build();

        System.out.println(dataFrame.head(10));

        TupleTwo<DataFrame, DataFrame> frames = dataFrame.shuffle().split(0.9);

        DataFrame trainingData = frames._1();
        System.out.println(trainingData.head(10));

        DataFrame crossValidationData = frames._2();

        MLPRegression regression = new MLPRegression();
        regression.setWeightUpdateMode(WeightUpdateMode.MiniBatchGradientDescend);
        regression.setMiniBatchSize(20);
        regression.setOutputNormalization(new RangeScaler());
        regression.setHiddenLayers(8);
        regression.setEpoches(1000);
        regression.fit(trainingData);

        for(int i = 0; i < crossValidationData.rowCount(); ++i){
            double predicted = regression.transform(crossValidationData.row(i));
            double actual = crossValidationData.row(i).target();
            logger.info("predicted: {}\texpected: {}", predicted, actual);
        }

`

**Comment by midikiman** [comment]
Hi, sorry for the late reply, I'll attach my source code below. From the code you posted, the problem may be in having both categorical and numeric columns in the input, unless there's something wrong with my other parameters.

            String columnSplitter = ",";
            boolean skipFirstLine = true;
            InputStream inputStream = new FileInputStream("c:/proj/transparent_aluminum/train.csv");
            DataFrame frame = DataQuery.csv(columnSplitter, skipFirstLine)
                    .from(inputStream)
                    .selectColumn(1).asCategory().asInput("spacegroup")
                    .selectColumn(2).asCategory().asInput("number_of_total_atoms")
                    .selectColumn(3).asNumeric().asInput("percent_atom_al")
                    .selectColumn(4).asNumeric().asInput("percent_atom_ga")
                    .selectColumn(5).asNumeric().asInput("percent_atom_in")
                    .selectColumn(12).asNumeric().asOutput("formation_energy_ev_natom")
                    .build();
            System.out.println(frame.rowCount());
            System.out.println(frame.getLevels());
            TupleTwo<DataFrame, DataFrame> miniFrames = frame.shuffle().split(0.9);

            DataFrame trainingData = miniFrames._1();
            DataFrame crossValidationData = miniFrames._2();

            MLPRegression regression = new MLPRegression();
            regression.setHiddenLayers(8);
            regression.setEpoches(1000);
            regression.fit(trainingData);
            
            System.out.println("Training:");
            for (int i = 0; i < trainingData.rowCount(); ++i) {
                double[] array = trainingData.row(i).toArray();
                double predicted = regression.transform(trainingData.row(i));
                double actual = trainingData.row(i).target();
                System.out.format("predicted: %f\texpected: %f\n", predicted, actual);
            }

            System.out.println("Cross-validation:");
            for (int i = 0; i < crossValidationData.rowCount(); ++i) {
                double predicted = regression.transform(crossValidationData.row(i));
                double actual = crossValidationData.row(i).target();
                System.out.format("predicted: %f\texpected: %f\n", predicted, actual);
            }


----- Original Message -----

**Comment by midikiman** [comment]
OK, it's the other parameters. When I added your settings for weightUpdateMode, miniBatchSize, and outputNormalization, it gives numbers instead of NaN. (They aren't great numbers, but that's not the fault of your library. :) )

----- Original Message -----
